src/esp/physics/bullet/BulletDebugManager.cpp

```diff
@@ -0,0 +1,186 @@
+    const btCollisionObject* colObj1 =
+        static_cast<const btCollisionObject*>(manifold->getBody1());
+
+    // logic copied from btSimulationIslandManager::buildIslands. We want to
+    // count manifolds only if related to non-sleeping bodies.
+    if (((colObj0) && colObj0->getActivationState() != ISLAND_SLEEPING) ||
+        ((colObj1) && colObj1->getActivationState() != ISLAND_SLEEPING)) {
+      func(colObj0, colObj1, manifold);
+    }
+  }
+}
+
+std::string BulletDebugManager::getDebugStringForCollisionObject(
+    const btCollisionObject* colObj) {
+  std::string name = (collisionObjectToDebugName_.count(colObj))
+                         ? collisionObjectToDebugName_[colObj]
+                         : "unknown";
+
+  // reference code to shorten names
+  // const int maxLen = 60;
+  // if (name.size() > maxLen) {
+  //  name = name.substr(name.size() - maxLen, maxLen);
+  //}
+
+  /*
+  #define ACTIVE_TAG 1
+  #define ISLAND_SLEEPING 2
+  #define WANTS_DEACTIVATION 3
+  #define DISABLE_DEACTIVATION 4
+  #define DISABLE_SIMULATION 5
+  */
+  const char* activationStateNames[] = {"invalid",
+                                        "active",
+                                        "sleeping",
+                                        "wants_deactivation",
+                                        "disable_deactivation",
+                                        "disable_simulation"};
+  std::stringstream result;
+
+  result << static_cast<const void*>(colObj) << " " << name << ", "
+         << activationStateNames[colObj->getActivationState()];
+
+  const auto* broadphaseHandle = colObj->getBroadphaseHandle();
+  if (broadphaseHandle) {
+    result << ", group "
+           << std::to_string(broadphaseHandle->m_collisionFilterGroup) +
+                  ", mask "
+           << std::to_string(broadphaseHandle->m_collisionFilterMask);
+  } else {
+    // this probably means the object hasn't been added to the bullet world
+    result << ", no broadphase handle";
+  }
+  return result.str();
+}
+
+std::string BulletDebugManager::getStepCollisionSummary(
+    btMultiBodyDynamicsWorld* bWorld) {
+  std::stringstream s;
+  auto func = [&](const btCollisionObject* colObj0,
+                  const btCollisionObject* colObj1,
+                  const btPersistentManifold* manifold) {
+    if (manifold->getNumContacts() > 0) {
+      s << "[" << getDebugStringForCollisionObject(colObj0) << "] vs ["
+        << getDebugStringForCollisionObject(colObj1) << "], "
+        << manifold->getNumContacts() << " points" << std::endl;
+    }
+  };
+
+  processActiveManifolds(bWorld, func);
+
+  const bool isEmpty = s.rdbuf()->in_avail() == 0;
+  if (isEmpty) {
+    s << "(no active collision manifolds)" << std::endl;
+  }
+
+  return s.str();
+}
+
+std::string BulletDebugManager::getCollisionFilteringSummary(bool doVerbose) {
+  std::stringstream s2;
+  for (const auto& pair0 : collisionObjectToDebugName_) {
```

Comment:
The declaration for the function mentions how this function is unsafe (it can crash). The issue is that Bullet collision objects can be deleted but this debug manager doesn't get notified of that.

One solution is to change this function to start by iterating through all Bullet collision objects known to Bullet (this list must exist somewhere). In this way, we naturally skip over items in collisionObjectToDebugName_ corresponding to deleted collision objects.